### PR TITLE
Fetch BIDS information from schema

### PR DESCRIPTION
**Pull request by nx10**
This PR makes bids2table fetch BIDS entity definitions and BIDS datatypes from the official BIDS schema.
This will ensure bids2table is always up-to-date and does not need to be changed when the schema changes.

Custom schema definitions can be injected via `bids2table.entities.set_bids_schema_path(...)`.

See also https://github.com/ohbm/hackathon2024/issues/22

@effigies

~Note: `run` (incorrectly) was an integer before - hence certain changes in the tests.~

### CODE REVIEW

File: bids2table/entities.py [hunk 1]
```diff
@@ -257,7 +222,34 @@ def with_update(
...
+        "name": entity_schema["name"],
+        "display_name": entity_schema["display_name"],
+        "allowed_values": entity_schema.get("enum"),
+    }
+    type_ = {"integer": int}.get(entity_schema["type"], str)
+    field_ = field(default=None, metadata=metadata)
+    return entity_schema["name"], Optional[type_], field_
+
+
+BIDSEntities: Any = make_dataclass(
```
Comment (effigies) [inline bids2table/entities.py:244]:
```suggestion
BIDSEntities: type = make_dataclass(
```

File: bids2table/entities.py [hunk 2]
```diff
@@ -248,7 +213,7 @@ def to_path(
 
     def with_update(
         self, entitities: Optional[Dict[str, Any]] = None, **kwargs
-    ) -> "BIDSEntities":
+    ) -> "_BIDSEntitiesBase":
```
Comment (effigies) [inline bids2table/entities.py:216]: You can use `Self` to indicate that the output has the same type as input.

```suggestion
    ) -> Self:
```

File: bids2table/entities.py [hunk 3]
```diff
@@ -4,28 +4,38 @@
 
 import re
 import warnings
-from dataclasses import asdict, dataclass, field, fields
+from dataclasses import asdict, dataclass, field, fields, make_dataclass
 from functools import lru_cache
 from pathlib import Path
 from types import MappingProxyType
-from typing import Any, Callable, Dict, Iterable, List, Optional, Union
+from typing import Any, Callable, Dict, Iterable, List, Optional, Tuple, Union
```
Comment (effigies) [inline bids2table/entities.py:11]:
```suggestion
from typing import Any, Callable, Dict, Iterable, List, Optional, Self, Tuple, Union
```

File: bids2table/entities.py [hunk 4]
```diff
@@ -4,28 +4,38 @@
...
+def _get_bids_schema() -> Any:
+    """
+    Get the BIDS schema.
+    """
+    global _bids_schema_path
+    return bidsschematools.schema.load_schema(schema_path=_bids_schema_path)
+
+
+BIDS_DATATYPES = tuple(
+    o.value for o in dict(_get_bids_schema().objects.datatypes).values()
```
Comment (effigies) [inline bids2table/entities.py:38]: This is already a dict-like:

```suggestion
    o.value for o in _get_bids_schema().objects.datatypes.values()
```

File: bids2table/entities.py [hunk 5]
```diff
@@ -257,7 +222,34 @@ def with_update(
...
+) -> Tuple[str, Any, Any]:
+    """
+    BIDS entity dataclass field.
+    """
+    metadata = {
+        "name": entity_schema["name"],
+        "display_name": entity_schema["display_name"],
+        "allowed_values": entity_schema.get("enum"),
+    }
+    type_ = {"integer": int}.get(entity_schema["type"], str)
```
Comment (effigies) [inline bids2table/entities.py:239]: So they're all strings, but `format: index` indicates that they should be interpreted as integers:

```suggestion
    type_ = {"index": int}.get(entity_schema["format"], str)
```

File: bids2table/entities.py (hunk 1 above)
Comment (nx10) [reply bids2table/entities.py:248]: I had it like this, but mypy does not like it at all (results in everything being flagged missing-attr). There is an open issue: https://github.com/python/mypy/issues/6063

Comment (effigies) [reply bids2table/entities.py:248]: Cool...

File: tests/test_table.py [hunk 6]
```diff
@@ -69,7 +70,7 @@ def test_table_files(tab: BIDSTable):
     }
 
     ents = file.entities
-    assert (ents.sub, ents.task, ents.run) == ("01", "balloonanalogrisktask", 1)
+    assert (ents.sub, ents.task, ents.run) == ("01", "balloonanalogrisktask", "01")
```
Comment (effigies) [inline tests/test_table.py:73]:
```suggestion
    assert (ents.sub, ents.task, ents.run) == ("01", "balloonanalogrisktask", 1)
```

File: tests/test_entities.py [hunk 7]
```diff
@@ -24,7 +24,7 @@
             sub="A01",
             ses="B02",
             task="rest",
-            run=1,
+            run="1",
```
Comment (effigies) [inline tests/test_entities.py:27]:
```suggestion
            run=1,
```

**Review by effigies — COMMENTED** [review]
I think these need to be reverted now?

**Comment by clane9** [comment]
This is a very nice addition, thanks @nx10, @effigies! I think it looks good to me.

**Pull request merged**